components/script/dom/headers.rs

```diff
@@ -469,3 +467,69 @@ pub fn is_obs_text(x: u8) -> bool {
         _ => false,
     }
 }
+
+// https://fetch.spec.whatwg.org/#concept-header-extract-mime-type
+// This function uses data_url::Mime to parse the MIME Type because
+// mime::Mime does not provide a parser following the Fetch spec
+// see https://github.com/hyperium/mime/issues/106
+pub fn extract_mime_type(headers: &HyperHeaders) -> Option<Vec<u8>> {
+    let mut charset: Option<String> = None;
+    let mut essence: String = "".to_string();
+    let mut mime_type: Option<DataUrlMime> = None;
+
+    // Step 4
+    let headers_values = headers.get_all(http::header::CONTENT_TYPE).iter();
+
+    // Step 5
+    if headers_values.size_hint() == (0, Some(0)) {
+        return None;
+    }
+
+    // Step 6
+    for header_value in headers_values {
+        // Step 6.1
+        match DataUrlMime::from_str(header_value.to_str().unwrap_or("")) {
+            // Step 6.2
+            Err(_) => continue,
+            Ok(temp_mime) => {
+                let temp_essence = format!("{}/{}", temp_mime.type_, temp_mime.subtype);
+
+                // Step 6.2
+                if temp_essence == "*/*" {
+                    continue;
+                }
+
+                let temp_charset = &temp_mime.get_parameter(&"charset".to_string());
```

Comment:
The to_string shouldn't be necessary, I believe.

Reply:
You're right! Fixed